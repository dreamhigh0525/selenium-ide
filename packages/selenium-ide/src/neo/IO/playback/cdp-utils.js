// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

export function buildFrameTree(docTree) {
  const doc = {
    children: parseFramesFromDoc(docTree),
  }
  return doc
}

function parseFramesFromDoc(docTree) {
  const elements = docTree.children ? [...docTree.children] : []
  const frames = []

  while (elements.length) {
    const elem = elements.pop()
    if (elem.nodeName === 'IFRAME') {
      frames.push(elem)
    } else if (elem.children) {
      elements.push(...elem.children)
    }
  }

  return frames
    .map(frame => ({
      nodeId: frame.nodeId,
      backendNodeId: frame.backendNodeId,
      parentId: frame.parentId,
      frameId: frame.frameId,
      documentURL: getUrlForFrame(frame),
      documentNodeId: getDocumentNodeIdForFrame(frame),
      children: frame.contentDocument
        ? parseFramesFromDoc(frame.contentDocument)
        : [],
    }))
    .sort((frm1, frm2) => frm1.nodeId - frm2.nodeId)
}

function getUrlForFrame(frame) {
  const srcAttrIndex = frame.attributes.indexOf('src')
  return srcAttrIndex !== -1 ? frame.attributes[srcAttrIndex + 1] : undefined
}

function getDocumentNodeIdForFrame(frame) {
  try {
    return frame.contentDocument.children.find(
      child =>
        child.children &&
        child.children.length &&
        child.children.length > 1 &&
        child.children[1].nodeName === 'BODY'
    ).children[1].nodeId
  } catch (e) {
    return undefined
  }
}
