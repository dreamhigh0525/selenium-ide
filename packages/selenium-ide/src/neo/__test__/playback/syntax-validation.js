// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import { validateControlFlowSyntax } from "../../playback/playback-tree/syntax-validation";
import Command, { ControlFlowCommandNames } from "../../models/Command";

function createCommand(name) {
  return new Command(null, name, "", "");
}

describe("Control Flow", () => {
  describe("Preprocess", () => {
    describe("Syntax Validation", () => {
      test("if, end", () => {
        let result = validateControlFlowSyntax([
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.end)
        ]);
        expect(result).toBeTruthy();
      });
      test("if, else, end", () => {
        let result = validateControlFlowSyntax([
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.else),
          createCommand(ControlFlowCommandNames.end)
        ]);
        expect(result).toBeTruthy();
      });
      test("if, elseIf, end", () => {
        let result = validateControlFlowSyntax([
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.elseIf),
          createCommand(ControlFlowCommandNames.end)
        ]);
        expect(result).toBeTruthy();
      });
      test("if, elseIf, else, end", () => {
        let result = validateControlFlowSyntax([
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.elseIf),
          createCommand(ControlFlowCommandNames.else),
          createCommand(ControlFlowCommandNames.end)
        ]);
        expect(result).toBeTruthy();
      });
      test("while, end", () => {
        let result = new validateControlFlowSyntax([
          createCommand(ControlFlowCommandNames.while),
          createCommand(ControlFlowCommandNames.end)
        ]);
        expect(result).toBeTruthy();
      });
      test("times, end", () => {
        let result = validateControlFlowSyntax([
          createCommand("times"),
          createCommand(ControlFlowCommandNames.end)
        ]);
        expect(result).toBeTruthy();
      });
      test("do, repeatIf", () => {
        let result = validateControlFlowSyntax([
          createCommand(ControlFlowCommandNames.do),
          createCommand(ControlFlowCommandNames.repeatIf)
        ]);
        expect(result).toBeTruthy();
      });
      test("do, while, end, repeatIf", () => {
        let result = validateControlFlowSyntax([
          createCommand(ControlFlowCommandNames.do),
          createCommand(ControlFlowCommandNames.while),
          createCommand(ControlFlowCommandNames.end),
          createCommand(ControlFlowCommandNames.repeatIf)
        ]);
        expect(result).toBeTruthy();
      });
    });
    describe("Syntax Invalidation", () => {
      test("if", () => {
        let input = [createCommand(ControlFlowCommandNames.if)];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("Incomplete block at if");
      });
      test("if, if, end", () => {
        let input = [
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.end)
        ];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("Incomplete block at if");
      });
      test("if, else, elseIf, end", () => {
        let input = [
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.else),
          createCommand(ControlFlowCommandNames.elseIf),
          createCommand(ControlFlowCommandNames.end)
        ];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("Incorrect command order of else if / else");
      });
      test("if, else, else, end", () => {
        let input = [
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.else),
          createCommand(ControlFlowCommandNames.else),
          createCommand(ControlFlowCommandNames.end)
        ];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("Too many else commands used");
      });
      test("else", () => {
        let input = [ createCommand(ControlFlowCommandNames.else) ];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("An else used outside of an if block");
      });
      test("elseIf", () => {
        let input = [ createCommand(ControlFlowCommandNames.elseIf) ];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("An else if used outside of an if block");
      });
      test("while", () => {
        let input = [createCommand(ControlFlowCommandNames.while)];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("Incomplete block at while");
      });
      test("if, while", () => {
        let input = [
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.while)
        ];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("Incomplete block at while");
      });
      test("if, while, end", () => {
        let input = [
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.while),
          createCommand(ControlFlowCommandNames.end)
        ];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("Incomplete block at if");
      });
      test("if, while, else, end", () => {
        let input = [
          createCommand(ControlFlowCommandNames.if),
          createCommand(ControlFlowCommandNames.while),
          createCommand(ControlFlowCommandNames.else),
          createCommand(ControlFlowCommandNames.end)
        ];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("An else used outside of an if block");
      });
      test("times", () => {
        let input = [createCommand("times")];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("Incomplete block at times");
      });
      test("repeatIf", () => {
        let input = [createCommand(ControlFlowCommandNames.repeatIf)];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("A repeat if used without a do block");
      });
      test("do", () => {
        let input = [createCommand(ControlFlowCommandNames.do)];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("Incomplete block at do");
      });
      test("end", () => {
        let input = [createCommand(ControlFlowCommandNames.end)];
        expect(function() { validateControlFlowSyntax(input); }).toThrow("Use of end without an opening keyword");
      });
    });
  });
});
