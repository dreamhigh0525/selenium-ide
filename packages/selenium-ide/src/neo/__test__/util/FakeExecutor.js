// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

export default class FakeExecutor {
  async init({ _baseUrl, _variables }) {}

  async cleanup() {}

  isWebDriverCommand() {
    return true
  }

  name(command) {
    if (!command) {
      return 'skip'
    }

    const upperCase = command.charAt(0).toUpperCase() + command.slice(1)
    const func = 'do' + upperCase
    if (!this[func]) {
      throw new Error(`Unknown command ${command}`)
    }
    return func
  }

  async beforeCommand(_commandObject) {}

  async afterCommand(_commandObject) {}

  async doPause(timeout = 0) {
    await new Promise(res => {
      setTimeout(res, parseInt(timeout))
    })
  }
}
