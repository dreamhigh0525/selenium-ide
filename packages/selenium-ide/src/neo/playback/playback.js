// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import EventEmitter from 'events'
import { createPlaybackTree } from './playback-tree'
import { mergeEventEmitter } from '../../common/events'
import { AssertionError, VerificationError } from './errors'

const EE = Symbol('event-emitter')

export default class Playback {
  constructor({ executor, testId, baseUrl, variables }) {
    this.executor = executor
    this.testId = testId
    this.baseUrl = baseUrl
    this.variables = variables
    this[EE] = new EventEmitter()
    mergeEventEmitter(this, this[EE])
  }

  async init() {
    await this._prepareToPlay()
    this._initialized = true
  }

  async play(commands) {
    this.playbackTree = createPlaybackTree(commands)
    this.currentExecutingNode = this.playbackTree.startingCommandNode
    if (!this._initialized) await this.init()
    await this._play()
  }

  async playSingleCommand() {}

  async pause() {}

  async resume() {}

  async stop() {
    this._setExitCondition(PlaybackStates.STOPPED)
    this._stopping = true
    await this.executor.cancel()

    // play will throw but the user will catch it with this.play()
    // this.stop() should resolve once play finishes
    await this._playPromise.catch(() => {})
  }

  async abort() {
    this._setExitCondition(PlaybackStates.ABORTED)
    await this.executor.kill()

    // play will throw but the user will catch it with this.play()
    // this.abort() should resolve once play finishes
    await this._playPromise.catch(() => {})
  }

  async cleanup() {
    await this.executor.cleanup()
  }

  async _prepareToPlay() {
    this[EE].emit(PlaybackEvents.PLAYBACK_STATE_CHANGED, {
      state: PlaybackStates.PREPARATION,
    })
    await this.executor.init({
      baseUrl: this.baseUrl,
      variables: this.variables,
    })
  }

  async _play() {
    this[EE].emit(PlaybackEvents.PLAYBACK_STATE_CHANGED, {
      state: PlaybackStates.PLAYING,
    })
    this._playPromise = (async () => {
      try {
        await this._executionLoop()
      } catch (err) {
        throw err
      } finally {
        await this._finishPlaying()
      }
    })()

    await this._playPromise
  }

  async _executionLoop() {
    if (this._stopping) {
      return
    } else if (this.currentExecutingNode) {
      const command = this.currentExecutingNode.command
      this[EE].emit(PlaybackEvents.COMMAND_STATE_CHANGED, {
        id: command.id,
        callstackIndex: undefined,
        state: CommandStates.Pending,
        message: undefined,
      })
      let result
      try {
        result = await this._executeCommand(this.currentExecutingNode)
      } catch (err) {
        if (err instanceof AssertionError) {
          this[EE].emit(PlaybackEvents.COMMAND_STATE_CHANGED, {
            id: command.id,
            callstackIndex: undefined,
            state: CommandStates.Failed,
            message: err.message,
          })
          this._setExitCondition(PlaybackStates.FAILED)
          throw err
        } else if (err instanceof VerificationError) {
          this[EE].emit(PlaybackEvents.COMMAND_STATE_CHANGED, {
            id: command.id,
            callstackIndex: undefined,
            state: CommandStates.Failed,
            message: err.message,
          })
          this._setExitCondition(PlaybackStates.FAILED)
          // focibly continuing verify commands
          this.currentExecutingNode = this.currentExecutingNode.next
          return await this._executionLoop()
        } else {
          this[EE].emit(PlaybackEvents.COMMAND_STATE_CHANGED, {
            id: command.id,
            callstackIndex: undefined,
            state: CommandStates.Fatal,
            message: err.message,
          })
          this._setExitCondition(PlaybackStates.ERRORED)
          throw err
        }
      }
      this[EE].emit(PlaybackEvents.COMMAND_STATE_CHANGED, {
        id: command.id,
        callstackIndex: undefined,
        state: CommandStates.Passed,
        message: undefined,
      })
      this.currentExecutingNode = result.next

      return await this._executionLoop()
    }
  }

  async _executeCommand(commandNode) {
    const result = await commandNode.execute(this.executor)

    return result
  }

  async _finishPlaying() {
    this[EE].emit(PlaybackEvents.PLAYBACK_STATE_CHANGED, {
      state: this._exitCondition || PlaybackStates.FINISHED,
    })
  }

  _setExitCondition(condition) {
    if (!this._exitCondition) {
      this._exitCondition = condition
    } else if (
      PlaybackStatesPriorities[condition] >
      PlaybackStatesPriorities[this._exitCondition]
    ) {
      this._exitCondition = condition
    }
  }
}

export const PlaybackEvents = {
  COMMAND_STATE_CHANGED: 'command-state-changed',
  PLAYBACK_STATE_CHANGED: 'playback-state-changed',
}

export const PlaybackStates = {
  PREPARATION: 'prep',
  PLAYING: 'playing',
  FINISHED: 'finished',
  FAILED: 'failed',
  ERRORED: 'errored',
  PAUSED: 'paused',
  BREAKPOINT: 'breakpoint',
  STOPPED: 'stopped',
  ABORTED: 'aborted',
}

const PlaybackStatesPriorities = {
  [PlaybackStates.FINISHED]: 0,
  [PlaybackStates.FAILED]: 1,
  [PlaybackStates.ERRORED]: 2,
  [PlaybackStates.STOPPED]: 3,
  [PlaybackStates.ABORTED]: 4,
}

export const CommandStates = {
  Failed: 'failed',
  Fatal: 'fatal',
  Passed: 'passed',
  Pending: 'pending',
  Undetermined: 'undetermined',
  Awaiting: 'awaiting',
}
