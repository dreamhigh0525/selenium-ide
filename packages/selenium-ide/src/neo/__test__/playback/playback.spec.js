// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Playback, {
  PlaybackEvents,
  PlaybackStates,
  CommandStates,
} from '../../playback/playback'
import { AssertionError, VerificationError } from '../../playback/errors'
import FakeExecutor from '../util/FakeExecutor'

describe('Playback', () => {
  describe('Playback test queue', () => {
    it('should play a test', async () => {
      const test = [
        {
          command: 'open',
          target: '',
          value: '',
        },
        {
          command: 'open',
          target: '',
          value: '',
        },
        {
          command: 'open',
          target: '',
          value: '',
        },
      ]
      const executor = new FakeExecutor({})
      executor.doOpen = jest.fn(async () => {})
      const playback = new Playback({
        executor,
      })
      await playback.play(test)
      expect(executor.doOpen).toHaveBeenCalledTimes(3)
    })

    it('should fail to play a test with an unknown command', async () => {
      const test = [
        {
          command: 'fail',
          target: '',
          value: '',
        },
      ]
      const executor = new FakeExecutor({})
      const playback = new Playback({
        executor,
      })
      return expect(playback.play(test)).rejects.toBeInstanceOf(Error)
    })

    it('should pass a test with a failing verify', async () => {
      const test = [
        {
          command: 'verifyText',
          target: '',
          value: '',
        },
      ]
      const executor = new FakeExecutor({})
      executor.doVerifyText = jest.fn(async () => {
        throw new VerificationError('failed to verify')
      })
      const playback = new Playback({
        executor,
      })
      expect(async () => await playback.play(test)).not.toThrow()
    })
  })

  describe('abort', () => {
    it('should abort a test', async () => {
      const test = [
        {
          command: 'open',
          target: '',
          value: '',
        },
        {
          command: 'open',
          target: '',
          value: '',
        },
        {
          command: 'open',
          target: '',
          value: '',
        },
      ]
      const executor = new FakeExecutor({})
      executor.doOpen = jest.fn(() => psetTimeout(10))
      const playback = new Playback({
        executor,
      })
      const cb = jest.fn()
      playback.on(PlaybackEvents.PLAYBACK_STATE_CHANGED, cb)
      const playbackPromise = playback.play(test).catch(() => {})

      await psetTimeout(15)
      await playback.abort()
      await playbackPromise

      const results = cb.mock.calls.flat()
      expect(results.length).toBe(3)
      expect(results[0].state).toBe(PlaybackStates.PREPARATION)
      expect(results[1].state).toBe(PlaybackStates.PLAYING)
      expect(results[2].state).toBe(PlaybackStates.ABORTED)
    })
  })

  describe('Events', () => {
    describe("'command-state-changed'", () => {
      it('should listen to pending and pass changes', async () => {
        const test = [
          {
            command: 'open',
            target: '',
            value: '',
          },
        ]
        const executor = new FakeExecutor({})
        executor.doOpen = jest.fn(async () => {})
        const playback = new Playback({
          executor,
        })
        const cb = jest.fn()
        playback.on(PlaybackEvents.COMMAND_STATE_CHANGED, cb)
        await playback.play(test)
        const results = cb.mock.calls.flat()
        expect(results.length).toBe(2)
        expect(results[0].state).toBe(CommandStates.Pending)
        expect(results[1].state).toBe(CommandStates.Passed)
      })
      it('should listen to fail changes', async () => {
        const test = [
          {
            command: 'verifyText',
            target: '',
            value: '',
          },
        ]
        const executor = new FakeExecutor({})
        executor.doVerifyText = jest.fn(async () => {
          throw new VerificationError('failed to verify')
        })
        const playback = new Playback({
          executor,
        })
        const cb = jest.fn()
        playback.on(PlaybackEvents.COMMAND_STATE_CHANGED, cb)
        await playback.play(test).catch(() => {})
        const results = cb.mock.calls.flat()
        expect(results.length).toBe(2)
        expect(results[0].state).toBe(CommandStates.Pending)
        expect(results[1].state).toBe(CommandStates.Failed)
      })
      it('should listen to fatal changes', async () => {
        const test = [
          {
            command: 'fatal',
            target: '',
            value: '',
          },
        ]
        const executor = new FakeExecutor({})
        const playback = new Playback({
          executor,
        })
        const cb = jest.fn()
        playback.on(PlaybackEvents.COMMAND_STATE_CHANGED, cb)
        await playback.play(test).catch(() => {})
        const results = cb.mock.calls.flat()
        expect(results.length).toBe(2)
        expect(results[0].state).toBe(CommandStates.Pending)
        expect(results[1].state).toBe(CommandStates.Fatal)
      })
    })
    describe("'playback-state-changed'", () => {
      it('should finish test successfully', async () => {
        const test = [
          {
            command: 'open',
            target: '',
            value: '',
          },
        ]
        const executor = new FakeExecutor({})
        executor.doOpen = jest.fn(async () => {})
        const playback = new Playback({
          executor,
        })
        const cb = jest.fn()
        playback.on(PlaybackEvents.PLAYBACK_STATE_CHANGED, cb)
        await playback.play(test)
        const results = cb.mock.calls.flat()
        expect(results.length).toBe(3)
        expect(results[0].state).toBe(PlaybackStates.PREPARATION)
        expect(results[1].state).toBe(PlaybackStates.PLAYING)
        expect(results[2].state).toBe(PlaybackStates.FINISHED)
      })
      it('should fail due to verify', async () => {
        const test = [
          {
            command: 'open',
            target: '',
            value: '',
          },
          {
            command: 'verifyText',
            target: '',
            value: '',
          },
          {
            command: 'open',
            target: '',
            value: '',
          },
        ]
        const executor = new FakeExecutor({})
        executor.doOpen = jest.fn(async () => {})
        executor.doVerifyText = jest.fn(async () => {
          throw new VerificationError('failed to verify')
        })
        const playback = new Playback({
          executor,
        })
        const cb = jest.fn()
        playback.on(PlaybackEvents.PLAYBACK_STATE_CHANGED, cb)
        await playback.play(test)
        const results = cb.mock.calls.flat()
        expect(results.length).toBe(3)
        expect(results[0].state).toBe(PlaybackStates.PREPARATION)
        expect(results[1].state).toBe(PlaybackStates.PLAYING)
        expect(results[2].state).toBe(PlaybackStates.FAILED)
      })
      it('should fail due to assertion', async () => {
        const test = [
          {
            command: 'open',
            target: '',
            value: '',
          },
          {
            command: 'assertText',
            target: '',
            value: '',
          },
          {
            command: 'open',
            target: '',
            value: '',
          },
        ]
        const executor = new FakeExecutor({})
        executor.doOpen = jest.fn(async () => {})
        executor.doAssertText = jest.fn(async () => {
          throw new AssertionError('failed to assert')
        })
        const playback = new Playback({
          executor,
        })
        const cb = jest.fn()
        playback.on(PlaybackEvents.PLAYBACK_STATE_CHANGED, cb)
        await playback.play(test).catch(() => {})
        const results = cb.mock.calls.flat()
        expect(results.length).toBe(3)
        expect(results[0].state).toBe(PlaybackStates.PREPARATION)
        expect(results[1].state).toBe(PlaybackStates.PLAYING)
        expect(results[2].state).toBe(PlaybackStates.FAILED)
      })
      it('should fail due to error', async () => {
        const test = [
          {
            command: 'open',
            target: '',
            value: '',
          },
          {
            command: 'assertText',
            target: '',
            value: '',
          },
          {
            command: 'open',
            target: '',
            value: '',
          },
        ]
        const executor = new FakeExecutor({})
        executor.doOpen = jest.fn(async () => {})
        executor.doAssertText = jest.fn(async () => {
          throw new Error('error in command')
        })
        const playback = new Playback({
          executor,
        })
        const cb = jest.fn()
        playback.on(PlaybackEvents.PLAYBACK_STATE_CHANGED, cb)
        await playback.play(test).catch(() => {})
        const results = cb.mock.calls.flat()
        expect(results.length).toBe(3)
        expect(results[0].state).toBe(PlaybackStates.PREPARATION)
        expect(results[1].state).toBe(PlaybackStates.PLAYING)
        expect(results[2].state).toBe(PlaybackStates.ERRORED)
      })
    })
  })
})

const psetTimeout = timeout =>
  new Promise(res => {
    setTimeout(res, timeout)
  })
