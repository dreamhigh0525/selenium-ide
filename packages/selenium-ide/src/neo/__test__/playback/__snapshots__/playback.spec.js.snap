// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playback Playback test queue should play a nested test case 1`] = `
Array [
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 1,
    "id": 4,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 1,
    "id": 4,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 1,
    "id": 5,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 1,
    "id": 5,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 1,
    "id": 6,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 1,
    "id": 6,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "passed",
  },
]
`;

exports[`Playback abort should abort a paused test 1`] = `
Array [
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": "playback is dead",
    "state": "errored",
  },
]
`;

exports[`Playback abort should abort a test 1`] = `
Array [
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": "playback is dead",
    "state": "errored",
  },
]
`;

exports[`Playback abort should abort a test after attempting to stop 1`] = `
Array [
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": "playback is dead",
    "state": "errored",
  },
]
`;

exports[`Playback pause on exceptions should pause for every type of error 1`] = `
Array [
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": "error",
    "state": "errored",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": "error in verify",
    "state": "failed",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": "error in assert",
    "state": "failed",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "passed",
  },
]
`;

exports[`Playback pause on exceptions should pause until the command is fixed 1`] = `
Array [
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": "error",
    "state": "errored",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": "error",
    "state": "errored",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "passed",
  },
]
`;

exports[`Playback resume resume after hitting a breakpoint 1`] = `
Array [
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "passed",
  },
]
`;

exports[`Playback resume should resume a paused test 1`] = `
Array [
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "passed",
  },
]
`;

exports[`Playback stop should stop a paused test 1`] = `
Array [
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": "Playback stopped",
    "state": "errored",
  },
]
`;

exports[`Playback stop should stop a test 1`] = `
Array [
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 1,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 2,
    "message": undefined,
    "state": "passed",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": undefined,
    "state": "executing",
  },
  Object {
    "callstackIndex": 0,
    "id": 3,
    "message": "Playback stopped",
    "state": "errored",
  },
]
`;
